internal/dnsconf/dnsconf.go

```diff
@@ -43,6 +44,17 @@ func newHTTPClientForDoH(beginning time.Time, handler model.Handler) *http.Clien
 	return &http.Client{Transport: transport}
 }
 
+func withPort(address, port string) string {
+	// Handle the case where port was not specified. We have written in
+	// a bunch of places that we can just pass a domain in this case and
+	// so we need to gracefully ensure this is still possible.
+	_, _, err := net.SplitHostPort(address)
+	if err != nil && strings.Contains(err.Error(), "missing port in address") {
+		address = net.JoinHostPort(address, port)
+	}
```

Comment:
This is a good question. I believe what I am doing is in line with [net.Dial documentation](https://golang.org/pkg/net/#Dial):

```Go
[...]
// For TCP and UDP networks, the address has the form "host:port".
// The host must be a literal IP address, or a host name that can be
// resolved to IP addresses.
// The port must be a literal port number or a service name.
// If the host is a literal IPv6 address it must be enclosed in square
// brackets, as in "[2001:db8::1]:80" or "[fe80::1%zone]:80".
// The zone specifies the scope of the literal IPv6 address as defined
// in RFC 4007.
// The functions JoinHostPort and SplitHostPort manipulate a pair of
// host and port in this form.
```

Because of the last sentence, I feel confident that, if an `host:port` pair is missing the port according to Go's own definition of missing a port in a `host:port` pair, then using Go's sanctioned function to add the port is fine. I deliberately have chosen to ignore any other possible error here and let `net.Dial` handle it.

Would it be more correct to parse the port, if a port is given (i.e. no error)? Yes, but I feel like this is redundant because we're about to pass the `address` to `net.Dial`, which will perform this check.

Would it be more correct to parse IP, in case of no error? This would fail if a domain name is passed, so it would restrict the input accepted by this function, which currently also includes domain names.

As I mentioned above, I structured the code knowing `net.Dial` is about to be invoked, therefore I didn't bother with checking all errors, but just wanted to patch `address` when a port is missing.